Add-Active-Storage,-Add-tests

fh0enix opened the pull request:
Description:

This pull request introduces the implementation of Active Storage for handling file attachments in the  model. Additionally, comprehensive tests have been added to ensure the correct functioning of the new features.Book
![image](https://github.com/fh0enix/books_reader/assets/13628239/1f6f8d99-2fe1-4332-b130-792b141d1f88)

Changes Made:

Active Storage Integration:

Added the necessary configuration for Active Storage in the  model.Book
Implemented the attachment of PDF files to  records using Active Storage.Book
Test Suite:

Added thorough tests to validate the behavior of Active Storage attachment.
Ensured that valid and invalid cases are covered to maintain code integrity.
Tested file uploads and their associations with the  model.Book
Files Modified:

app/models/book.rb
spec/models/book_spec.rb
spec/controllers/books_controller_spec.rb

Testing Instructions:
![image](https://github.com/fh0enix/books_reader/assets/13628239/ae836413-0caa-4cd4-a4ce-da59be6430f2)

Create a new Book with a valid PDF attachment.
Attempt to create a Book with invalid data, ensuring it fails gracefully.
Verify that PDF attachments are correctly associated with Book records in the database.

Please review these modifications thoroughly, paying special attention to the security aspects and the impact on user functionality. Your feedback and suggestions are highly appreciated to ensure the successful implementation of this critical update.

Reviewer:

@loqimean

fh0enix committed:
Added pdf field to books and updated the form

fh0enix committed:
The diff is about changing the method names from `collection` to `resourse` and `book_params` to `resourse_params` in the `BooksController` class.

fh0enix committed:
Add Active Storage tables

fh0enix committed:
Added a factory trait to attach a PDF to a book, and added a test for it. Also added a test for creating a book with a PDF in the controller.

fh0enix committed:
Changed the background color of the body from gray-700 to gray-400.

loqimean commented on app/assets/book_test/test.pdf:
стоп, а чому тестові файли зберігаються тут, а не в папці тестів??

fh0enix replied:
solved

loqimean commented on app/models/book.rb:
```diff
@@ -1,4 +1,6 @@
 class Book < ApplicationRecord
+  has_one_attached :pdf
```
а якщо я видалю книгу?

loqimean commented on app/views/books/_form.html.erb:
```diff
@@ -11,7 +11,10 @@
-     </div>
+    </div>
+    <div class="my-5">
+      <%= f.input :pdf, as: :file, label: false, hint: "PDF File (up to 20MB)" %>
```
а де та валідація на лесс зен 20мб?))

fh0enix replied:
solved

loqimean commented on spec/factories/books.rb:
```diff
@@ -4,5 +4,14 @@
     description { Faker::Lorem.paragraph }
+
+    trait :with_content do
+      after(:create) do |book|
```
а якщо я хочу зробити білд?

fh0enix replied:
solved

loqimean commented on spec/factories/books.rb:
```diff
@@ -4,5 +4,14 @@
+
+    trait :with_content do
+      after(:create) do |book|
+        content_path = Rails.root.join("app", "assets", "book_test", "test.pdf")
```
ой не там той файл має бути....

loqimean commented on spec/models/book_spec.rb:
```diff
@@ -6,4 +6,9 @@
   it { is_expected.to validate_length_of(:isbn).is_equal_to(13) }
+
+  it 'has content' do
+    book = FactoryBot.create(:book, :with_content)
```
ну а `let` для кого придумали?)

fh0enix replied:
solved

loqimean commented on spec/requests/books_spec.rb:
```diff
@@ -96,4 +96,27 @@
   end
+
+  describe "POST #create" do
+    context "with valid params" do
+      it "creates a new Book with an attached PDF" do
+        expect {
+          post :create, params: { book: valid_book_attributes.merge(pdf: fixture_file_upload('app/assets/book_test/test.pdf', 'application/pdf')) }
+        }.to change(Book, :count).by(1)
+        expect(assigns(:book).pdf).to be_attached
+        expect(response).to redirect_to(Book.last)
+      end
+    end
+
+    context "with invalid params" do
+      it "does not create a new Book" do
+        expect {
+          post :create, params: { book: invalid_book_attributes }
+        }.to_not change(Book, :count)
+        expect(assigns(:book)).to be_a_new(Book)
+        expect(response).to render_template(:new)
```
щось взагалі не гуд, бо:
1. чому багатолінійний блок написаний фігурними скобками?
2. чому ти не тестує те, що от ця книга стала доступна на сторінці?

fh0enix replied:
solved

fh0enix committed:
Update book model with pdf attachment validation and add tests

fh0enix committed:
🦄 refactor:

 Update paths to use `spec/fixtures` directory

fh0enix committed:
🧪 test:

 Update book creation spec to use a factory with content and test that the PDF is attached

fh0enix committed:
Merge branch 'Add-Devise-authentication,-Users-model,-Add-tests' of https://github.com/fh0enix/books_reader into Add-Active-Storage,-Add-tests

loqimean commented on app/models/book.rb:
```diff
@@ -1,8 +1,25 @@
+
   scope :ordered, -> { order(:title) }
+
+  def pdf_file_size
+    if pdf.attached? && pdf.byte_size > 20.megabytes
+      errors.add(:pdf, "file size must be less than 20MB")
+    end
+  end
+
+  def pdf_content_type
+    if pdf.attached? && !pdf.content_type.in?(["application/pdf"])
+      errors.add(:pdf, "must be a PDF file")
+    end
+  end
```
та наче має бути вбудований спосіб валідації розміру

fh0enix replied:
з коробки не знайшов, але є гем який надає такий функціонал => gem 'active_storage_validations'

loqimean replied:
ну знач його юзай)

fh0enix replied:
solved

loqimean commented on spec/factories/books.rb:
```diff
@@ -4,5 +4,19 @@
+
+    trait :with_content do
+      transient do
+        content_path { Rails.root.join("spec", "fixtures", "files", "test.pdf") }
```
spec/support/files

fh0enix replied:
fixed

loqimean commented on spec/factories/books.rb:
```diff
@@ -4,5 +4,19 @@
+        content_path { Rails.root.join("spec", "fixtures", "files", "test.pdf") }
+      end
+
+      after(:create) do |book, evaluator|
+        if book.persisted?
+          book.pdf.attach(io: File.open(evaluator.content_path),
+                          filename: "test.pdf",
+                          content_type: "application/pdf")
+        end
```
а чому після створення? давай відразу

fh0enix replied:
fixed

loqimean commented on spec/models/book_spec.rb:
```diff
@@ -1,9 +1,22 @@
+
+  it "validates size of attached file" do
+    expect(book.pdf).to be_attached
+    expect(book.pdf.byte_size).to be < 20.megabytes
```
а воно взагалі так працює?))

fh0enix replied:
Ппрацює ....
![image](https://github.com/fh0enix/books_reader/assets/13628239/5dbeddbf-9166-4b15-bd1c-4a91b9a24c97)

loqimean replied:
мене турбує не отримання розміру, а те, що ти робиш після експекта)

fh0enix replied:
solved

fh0enix committed:
Update `spec/factories/books.rb` and `spec/requests/books_spec.rb` to use `fixture_file_upload` from the `rack-test` gem instead of `Rails.root.join`.
Update `spec/fixtures/files/test.pdf` to be an empty file.

fh0enix committed:
Update book model to use ActiveStorage validations

fh0enix committed:
```
Update book_spec.rb

The diff removes an expectation that the size of the attached PDF file is not within 1 megabyte of 20 megabytes.
